README.md
bytewise
========

A binary serialization which sorts bytewise for arbitrarily complex data structures, respecting [typewise](https://github.com/deanlandolt/typewise) structured sorting efficiently.

This library defines a total order of possible data structures allowed in a keyspace and a binary encoding which sorts bytewise in this order. The ordering is a superset of both the sorting algorithm defined by [IndexedDB](http://www.w3.org/TR/IndexedDB/#key-construct) and the one defined by [CouchDB](http://wiki.apache.org/couchdb/View_collation). This serialization makes it easy to take advantage of the benefits of structural indexing on systems with fast but naïve binary indexing.


## Order of Supported Structures

This is the top level order of the various structures that may be encoded:

* `null`
* `false`
* `true`
* `Number` (numeric)
* `Date` (numeric, epoch offset)
* `Buffer` (bitwise)
* `String` (lexicographic)
* `Set` (componentwise with elements sorted)
* `Array` (componentwise)
* `Object` (componentwise string-keyed key/value pairs)
* `Map` (componentwise key/value pairs)
* `RegExp` (stringified lexicographic)
* `Function` (stringified lexicographic)
* `undefined`


These specific structures can be used to serialize the vast majority of javascript values in a way that can be sorted in an efficient, complete and sensible manner. Each value is prefixed with a type tag, and we do some bit munging to encode our values in such a way as to carefully preserve the desired sort behavior, even in the precense of structural nested.

For example, negative numbers are stored as a different *type* from positive numbers, with its sign bit stripped and its bytes inverted to ensure numbers with a larger magnitude come first. `Infinity` and `-Infinity` can also be encoded -- they are *nullary* types, encoded using just their type tag., same with `null` and `undefined`, and the boolean values `false`, `true`, . `Date` instances are stored just like `Number` instances, but as in IndexedDB, `Date` sorts after `Number` (including `Infinity`). `Buffer` data can be stored in the raw, and is sorted before `String` data. Then come the collection types -- `Array`, `Object`, along with the additional types defined by es6: `Map` and `Set`. We can even serialize `Function` values, reviving them in an isolated [Secure ECMAScript](https://code.google.com/p/es-lab/wiki/SecureEcmaScript) context where they can't do anything but calculate (if you have some optional dependencies installed).


## Unsupported Structures

This serialization accomodates a wide range of javascript structures, but it is not exhaustive. Objects or arrays with reference cycles, for instance, cannot be serialized. `NaN` is also illegal anywhere in a serialized value, as its presense is very likely indicative of an error. Moreover, sorting for `NaN` is completely nonsensical. (Similarly we may want to reject objects which are instances of `Error`.) Invalid `Date` objects are also illegal. Since `WeakMap` objects cannot be enumerated they are also illegal. Attempts to serialize any values which include these structures will throw a `TypeError`.


## Usage

`encode` serializes any supported type and returns a buffer, or throws if an unsupported structure is passed:
  
  ``` js
  var bytewise = require('bytewise');
  var assert = require('assert');
  // Helper to encode and then toString the buffer, defaults to binary but we'll use hex to show non-string values
  function encode(value, encoding) { return bytewise.encode(value).toString(encoding || 'binary') }

  // Many types can be respresented using only their type tag, a single byte
  // WARNING type tags are subject to change for the time being!
  assert.equal(encode(null), '\x10');
  assert.equal(encode(false), '\x20');
  assert.equal(encode(true), '\x21');
  assert.equal(encode(undefined), '\xf0');

  // Numbers are stored in 9 bytes -- 1 byte for the type tag and an 8 byte float
  assert.equal(encode(12345, 'hex'), '4240c81c8000000000');
  // Negative numbers are stored as positive numbers, but with a lower type tag and their bits inverted
  assert.equal(encode(-12345, 'hex'), '41bf37e37fffffffff');

  // All numbers, integer or floating point, are stored as IEEE 754 doubles
  assert.equal(encode(1.2345, 'hex'), '423ff3c083126e978d');
  assert.equal(encode(-1.2345, 'hex'), '41c00c3f7ced916872');

  // Serialization does not preserve the sign bit, so 0 is indistinguishable from -0
  assert.equal(encode(-0, 'hex'), '420000000000000000');
  assert.equal(encode(0, 'hex'), '420000000000000000');

  // We can even serialize Infinity and -Infinity, though we just use their type tag
  assert.equal(encode(-Infinity, 'hex'), '40');
  assert.equal(encode(Infinity, 'hex'), '43');

  // Dates are stored just like numbers, but with different (and higher) type tags
  assert.equal(encode(new Date(-12345), 'hex'), '51bf37e37fffffffff');
  assert.equal(encode(new Date(12345), 'hex'), '5240c81c8000000000');

  // Strings are encoded as utf8, prefixed with their type tag (0x70, or the "p" character)
  assert.equal(encode('foo'), 'pfoo');
  assert.equal(encode('föo'), 'pfÃ¶o');

  // Buffers are also left alone, other than being prefixed with their type tag (0x60)
  assert.equal(encode(new Buffer('ff00fe01', 'hex'), 'hex'), '60ff00fe01');

  // Arrays are just a series of values terminated with a null byte
  assert.equal(encode([ true, -1.2345 ], 'hex'), 'a02141c00c3f7ced91687200');

  // Strings are also legible when embedded in complex structures like arrays
  // Items in arrays are deliminted by null bytes, and a final end byte marks the end of the array
  assert.equal(encode([ 'foo' ]), '\xa0pfoo\x00\x00');

  // The 0x01 and 0xfe bytes are used to escape high and low bytes while preserving the correct collation
  assert.equal(encode([ new Buffer('ff00fe01', 'hex') ], 'hex'), 'a060fefe0101fefd01020000');

  // Complex types like arrays can be arbitrarily nested, and fixed-sized types don't require a terminating byte
  assert.equal(encode([ [ 'foo', true ], 'bar' ]), '\xa0\xa0\pfoo\x00\x21\x00\pbar\x00\x00');

  // Objects are just string-keyed maps, stored like arrays: [ k1, v1, k2, v2, ... ]
  assert.equal(encode({ foo: true, bar: 'baz' }), '\xb0pfoo\x00\x21\pbar\x00\pbaz\x00\x00');
  
  ```


`decode` parses a buffer and returns the structured data, or throws if malformed:
  
  ``` js
  var samples = [
    'foo √',
    null,
    '',
    new Date('2000-01-01T00:00:00Z'),
    42,
    undefined,
    [ undefined ],
    -1.1,
    {},
    [],
    true,
    { bar: 1 },
    [ { bar: 1 }, { bar: [ 'baz' ] } ],
    -Infinity,
    false
  ];
  var result = samples.map(bytewise.encode).map(bytewise.decode);
  assert.deepEqual(samples, result);
  ```


`compare` is just a convenience bytewise comparison function:

  ``` js
  var sorted = [
    null,
    false,
    true,
    -Infinity,
    -1.1,
    42,
    new Date('2000-01-01T00:00:00Z'),
    '',
    'foo √',
    [],
    [ { bar: 1 }, { bar: [ 'baz' ] } ],
    [ undefined ],
    {},
    { bar: 1 },
    undefined
  ];

  var result = samples.map(bytewise.encode).sort(bytewise.compare).map(bytewise.decode);
  assert.deepEqual(sorted, result);
  ```


## Use Cases

### Numeric indexing

This is surprisingly difficult to with vanilla LevelDB -- basic approaches require ugly hacks like left-padding numbers to make them sort lexicographically (which is prone to overflow problems). You could write a one-off comparator function in C, but there a number of drawbacks to this as well. This serializaiton solves this problem in a clean, generalizable way, taking advantage of properties of the byte sequences defined by the IEE 754 floating point standard.

### Namespaces, partitions and patterns

This is another really basic and oft-needed ammenity that isn't very easy out of the box in LevelDB. We reserve the lowest and highest bytes as abstract tags representing low and high key sentinals, allowing you to faithfully request all values in any portion of an array. Arrays can be used as namespaces without any leaky hacks, or even more detailed slicing can be done per element to implement wildcards or even more powerful pattern semantics for specific elements in the array keyspace.

### Document storage

It may be reasonably fast to encode and decode, but `JSON.stringify` is totally useless for storing objects as document records in a way that is of any use for range queries, where LevelDB and its ilk excel. Our serialization allows you to build indexes on top of your documents. Being unable to indexing purposes since it doesn't sort correctly. We fix that, and even expand on the range of serializable types available.

### Multilevel language-sensitive collation

You have a bunch of strings in a paritcular language-specific strings you want to index, but at the time of indexing you're not sure *how* sorted you need them. Queries may or may not care about case or punctionation differences, for instance. You can index your string as an array of weights, most-to-least specific, and prefixed by collation language (since our values are language-sensitive). There are [mechanisms available](http://www.unicode.org/reports/tr10/#Run-length_Compression) to compress this array to keep its size reasonable.

### Full-text search

Full-text indexing is a natural extension of the language-sensitive collation use case described above. Add a little lexing and stemming and basic full text search is close at hand. Structured indexes can be employed to make other more interesting search features possible as well.

### CouchDB-style "joins"

Build a view that colocates related subrecords, taking advantage of component-wise sorting of arrays to interleave them. This is a technique I first saw [employed by CouchDB](http://www.cmlenz.net/archives/2007/10/couchdb-joins). More recently [Akiban](http://www.akiban.com/) has formalized this concept of [table grouping](http://blog.akiban.com/how-does-table-grouping-compare-to-sql-server-indexed-views/) and brought it the SQL world. Our collation extends naturally to their idea of [hierarchical keys](http://blog.akiban.com/introducing-hkey/).

### Emulating other systems

Clients that wish to employ a subset of the full range of possible types above can preprocess values to coerce them into the desired simpler forms before serializing. For instance, if you were to build CouchDB-style indexing you could round-trip values through a `JSON` encode cycle (to get just the subset of types supported by CouchDB) before passing to `encode`, resulting in a collation that is identical to CouchDB. Emulating IndexedDB's collation would require preprocessing away `Buffer` data and `undefined` values. (TODO what else? Does it normalize `-0` values to `0`?)

### Embedding in the browser

While this particular serialization is only useful for binary indexing, the collation defined can easily be extended to embed inside indexeddb. At the top level all values would be arrays with the type tag as the first value. Any binary data would have be transcoded to a string in a manner that preserves sort order. Otherwise we can lean on IndexedDB's default sort behavior as much as possible.


## License

[MIT](http://deanlandolt.mit-license.org/)


CHANGELOG.md
# CHANGELOG


## 0.6.0

- refactor flat escape (for strings, buffers) to avoid bitshifting (more debuggable, should be faster too)

## 0.5.0

- use 0.8 version of es6-shim, adapt iteration to accept it's broken iterator protocol
- add Object type, refactor to avoid using Map unless explicitly passed in

## 0.4.0

... TODO

test/index.js
'use strict';

var bytewise = require('../bytewise');

var samples = {
  empty: [
    void 0,
    null
  ],
  boolean: [
    false,
    true
  ],
  number: [
    Number.NEGATIVE_INFINITY,
    -Number.MAX_INTEGER,
    -1000,
    -100,
    -10,
    -3,
    -2,
    -1,
    -0.1,
    -0.01,
    -0.001,
    -Number.EPSILON * 2,
    -Number.EPSILON,
    -0,
    0,
    Number.EPSILON,
    Number.EPSILON * 2,
    0.001,
    0.01,
    0.1,
    1,
    2,
    3,
    10,
    100,
    1000,
    Number.MAX_INTEGER,
    Number.POSITIVE_INFINITY
  ],
  date: [
    new Date(-1),
    new Date(-0),
    new Date(0),
    new Date(1),
    new Date('2001-09-09T01:46:39.999Z'),
    new Date(1000000000000),
    new Date('2001-09-09T01:46:40.001Z')
  ],
  buffer: [
    new Buffer([]),
    new Buffer([ 0 ]),
    new Buffer([ 0, 0 ]),
    new Buffer([ 0, 1 ]),
    new Buffer([ 1, 0 ]),
    new Buffer([ 1, 1 ]),
    new Buffer([ 255 ]),
    new Buffer([ 255, 0 ]),
    new Buffer([ 255, 255 ])
  ],
  string: [
    '',
    '\x00',
    '\x00\x00',
    '\x00\x01',
    '\x00\xff',
    'AA',
    'AB',
    'Aa',
    'Ab',
    'a',
    'aa',
    'ab',
    'b',
    'ba',
    'bb',
    '\xff',
    '\xff\x00',
    '\xff\xfe',
    '\xff\xff'
  ]
};

function shuffle(o) {
  for (var j, x, i = o.length; i; j = parseInt(Math.random() * i, 10), x = o[--i], o[i] = o[j], o[j] = x);
  return o;
}

var assert = require('assert');
function listEqual(a, b) {
  // TODO stringify functions
  return assert.deepEqual(a, b);
}

function getSample() {
  var sample = [];
  Object.keys(samples).forEach(function(key) {
    sample = sample.concat(samples[key]);
  });
  return sample;
}


var sample = getSample();
var shuffled = shuffle(sample.slice());
listEqual(sample, shuffled.map(bytewise.encode).sort(bytewise.compare).map(bytewise.decode));


function getArraySample(depth) {
  var sample = getSample();
  if (!depth) return sample;
  sample.forEach(function(item) {
    sample.push([ item ]);
  });
  sample.push([ getArraySample(depth - 1) ]);
  return sample;
}

var sample = getArraySample(2);
var shuffled = shuffle(sample.slice());
listEqual(sample, shuffled.map(bytewise.encode).sort(bytewise.compare).map(bytewise.decode));

// TODO we need specific array samples with mixed
// We also need tests with specific samples for objects and other maps, sets, and functions
// We need to expand on the string samples with multibyte utf-8 and characters outside BMP


bench.js
'use strict';
var bytewise = require('../bytewise');

// This is just a simple little benchmark to compare against JSON
function run(name, fn, value, total) {
  var start = Date.now();
  while (total--) {
    fn(value);
  }
  var time = Date.now() - start;
  console.log(name, ':', time, 'ms');
  return time;
}

// The case we'll want to optimize for is (possibly nested) arrays of primitives
var sample = [ 'foo', 123, [ 'bar', -4.56 ] ];

exports.all = function(total) {
  console.log(total + ' runs each:');
  run('JSON.stringify', JSON.stringify, sample, total);
  run('bytewise.encode', bytewise.encode, sample, total);

  run('JSON.parse', JSON.parse, JSON.stringify(sample), total);
  run('bytewise.decode', bytewise.decode, bytewise.encode(sample), total);
};

if (require.main === module) exports.all(10000);


bytewise.js
'use strict';
require('es6-shim');
// FIXME fix iterating keys in --harmony maps and sets

var compare = function(a, b) {
  var result;
  for (var i = 0, end = Math.min(a.length, b.length); i < end; i++) {
    result = a.get(i) - b.get(i);
    if (result) return result;
  }
  return a.length - b.length;
};

var _type = {
  function: {
    parse: function() {
      throw new Error('Fallback for function reviving NYI');
    },
    serialize: function() {
      throw new Error('Fallback for function serializing NYI');
    }
  }
};

try {
  var typewise = require('typewise');
  compare = typewise.comparators.bytewise;
  _type.function = typewise.types.function;
}
catch (e) {}

// Sort tags used to preserve binary total order
// The tag is 1 byte, which gives us plenty of room to grow.
// We leave some space between the various types for possible future compatibility with extensions.

// 0x00 reserved for termination character
var NULL = 0x10;
var FALSE = 0x20;
var TRUE = 0x21;
var NEGATIVE_INFINITY = 0x40;
var NEGATIVE_NUMBER = 0x41; // packed in an inverted form to sort bitwise ascending
var POSITIVE_NUMBER = 0x42;
var POSITIVE_INFINITY = 0x43;
var DATE_PRE_EPOCH = 0x51; // packed identically to a NEGATIVE_NUMBER
var DATE_POST_EPOCH = 0x52; // packed identically to a POSITIVE_NUMBER
var BUFFER = 0x60;
var STRING = 0x70;
var SET = 0x90; // packed as array with members sorted and deduped
var ARRAY = 0xa0; // escapes nested types with bit shifting where necessary to maintain order
var OBJECT = 0xb0; // just like couchdb member order is preserved and matters for collation
var MAP = 0xc0; // just like couchdb member order is preserved and matters for collation
var REGEXP = 0xd0; // packed as tuple of two strings, the end being flags
var FUNCTION = 0xe0; // packed as array, revived by safe eval in an isolated environment (if available)
var UNDEFINED = 0xf0;
// 0xff reserved for high-key sentinal


var flatTypes = [ BUFFER, STRING ];
var structuredTypes = [ ARRAY, OBJECT, MAP, SET, FUNCTION ];
var nullaryTypes = [ NULL, FALSE, TRUE, NEGATIVE_INFINITY, POSITIVE_INFINITY, UNDEFINED ];
var fixedTypes = {};
fixedTypes[NEGATIVE_NUMBER] = 8;
fixedTypes[POSITIVE_NUMBER] = 8;
fixedTypes[DATE_PRE_EPOCH] = 8;
fixedTypes[DATE_POST_EPOCH] = 8;


function encode(source) {

  if (source === void 0) return tag(UNDEFINED);
  if (source === null) return tag(NULL);

  // Unbox possible natives
  var value = source.valueOf();
  var type;

  // NaN and Invalid Date not permitted
  if (value !== value) {
    if (source instanceof Date) throw new TypeError('Invalid Date not permitted');
    throw new TypeError('NaN not permitted');
  }

  if (value === false) return tag(FALSE);
  if (value === true) return tag(TRUE);

  if (source instanceof Date) {
    // Normalize -0 values to 0
    if (Object.is(value, -0)) value = 0;
    type = value < 0 ? DATE_PRE_EPOCH : DATE_POST_EPOCH;
    return tag(type, encodeNumber(value));
  }

  if (typeof value === 'number') {
    if (value === Number.NEGATIVE_INFINITY) return tag(NEGATIVE_INFINITY);
    if (value === Number.POSITIVE_INFINITY) return tag(POSITIVE_INFINITY);
    // Normalize -0 values to 0
    if (Object.is(value, -0)) value = 0;
    type = value < 0 ? NEGATIVE_NUMBER : POSITIVE_NUMBER;
    return tag(type, encodeNumber(value));
  }

  // TODO also handle typed array, blob, etc.
  if (value instanceof Buffer) {
    return tag(BUFFER, value);
  }

  if (typeof value === 'string') {
    return tag(STRING, new Buffer(value, 'utf8'));
  }

  // RegExp
  if (value instanceof RegExp) {
    // TODO
    throw new Error('NYI');
  }

  // Function
  if (typeof value === 'function') {
    return tag(FUNCTION, encodeList(_type['function'].serialize(value)));
  }

  // Array
  // TODO better handling for sparse arrays
  if (Array.isArray(value)) {
    return tag(ARRAY, encodeList(value));
  }

  // Map
  if (value instanceof Map) {
    // Packs into an array, e.g. [ k1, v1, k2, v2, ... ]
    var items = [];
    getCollectionKeys(value).forEach(function(key) {
      items.push(key);
      items.push(value.get(key));
    });
    return tag(MAP, encodeList(items));
  }

  // Set
  if (value instanceof Set) {
    var set = getCollectionKeys(value);
    // encode, sort, and then decode the result array
    set = decode(set.map(encode).sort(compare));
    // TODO we should be able to build a list by concatenating buffers -- bypass this decode/encodeList dance
    return tag(SET, encodeList(set));
  }

  // Object
  if (typeof value === 'object' && Object.prototype.toString.call(value) === '[object Object]') {
    // Packs into an array, e.g. [ k1, v1, k2, v2, ... ]
    var items = [];
    Object.keys(value).forEach(function(key) {
      items.push(key);
      items.push(value[key]);
    });
    return tag(OBJECT, encodeList(items));
  }

  // TODO RegExp and other types from Structured Clone algorithm (Blob, File, FileList)

  throw new Error('Cannot encode unknown type: ' + source);
}

function decode(buffer) {

  var type = buffer[0];

  // Nullary types
  if (~nullaryTypes.indexOf(type)) {
    if (buffer.length !== 1) throw new Error('Invalid encoding in buffer: ' + buffer);

    if (type === UNDEFINED) return;
    if (type === NULL) return null;
    if (type === FALSE) return false;
    if (type === TRUE) return true;
    if (type === NEGATIVE_INFINITY) return Number.NEGATIVE_INFINITY;
    if (type === POSITIVE_INFINITY) return Number.POSITIVE_INFINITY;
  }

  // Fixed size types
  var chunk = buffer.slice(1);
  var chunkSize = fixedTypes[type];
  if (chunkSize) {
    if (chunk.length !== chunkSize) throw new Error('Invalid size for buffer: ' + buffer);

    if (type === NEGATIVE_NUMBER || type === POSITIVE_NUMBER) {
      return decodeNumber(chunk, type === NEGATIVE_NUMBER);
    }
    if (type === DATE_PRE_EPOCH || type === DATE_POST_EPOCH) {
      return new Date(decodeNumber(chunk, type === DATE_PRE_EPOCH));
    }
  }

  // Flat types
  if (type === BUFFER) return chunk;
  if (type === STRING) return chunk.toString('utf8');

  // Structured types
  if (~structuredTypes.indexOf(type)) {
    var result = parseHead(buffer);
    if (result[1] !== buffer.length) {
      throw new Error('List deserialization fail: ' + result[1] + '!=' + buffer.length);
    }
    return result[0];
  }

}


function tag(type, buffer) {
  // Just return tag byte for nullary types (no buffer provided)
  type = new Buffer([ type ]);
  if (!buffer) return type;
  // Prepend a type tag byte to buffer
  return Buffer.concat([ type, buffer ]);
}

function encodeNumber(value) {
  var buffer = new Buffer(8);
  if (value < 0) {
    buffer.writeDoubleBE(-value, 0);
    return invert(buffer);
  }
  buffer.writeDoubleBE(value, 0);
  return buffer;
}

function decodeNumber(buffer, negative) {
  if (negative) buffer = invert(buffer);
  var value = buffer.readDoubleBE(0);
  return negative ? -value : value;
}


function encodeList(items) {
  // TODO pass around a map of references already encoded to detect cycles
  var buffers = [];
  var chunk;
  for (var i = 0, end = items.length; i < end; ++i) {
    chunk = encode(items[i]);
    var type = chunk[0];
    // We need to escape a few bytes in string and buffer types to prevent confusion with the end byte
    if (~flatTypes.indexOf(type)) chunk = flatEscape(chunk);
    buffers.push(chunk);
  }
  // Close the list with an end byte
  buffers.push(new Buffer([ 0 ]));
  return Buffer.concat(buffers);
}

// TODO expose in public API
function flatEscape(buffer) {
  // Escape high and low bytes 0x00 and 0xff (and by necessity, 0x01 and 0xfe)
  var b, bytes = [];
  for (var i = 0, end = buffer.length; i < end; ++i) {
    b = buffer[i];
    // Escape low bytes with 0x01 and by adding 1
    if (b === 0x01 || b === 0x00) bytes.push(0x01, b + 1);
    // Escape high bytes with 0xfe and by subtracting 1
    else if (b === 0xfe || b === 0xff) bytes.push(0xfe, b - 1);
    // Otherwise no escapement needed
    else bytes.push(b);
  }
  // Add end byte
  bytes.push(0);
  return new Buffer(bytes);
}

// TODO expose in public API
function flatUnescape(buffer) {
  var b, bytes = [];
  // Don't escape last byte
  for (var i = 0, end = buffer.length; i < end; ++i) {
    b = buffer[i];
    // If low-byte escape tag use the following byte minus 1
    if (b === 0x01) bytes.push(buffer[++i] - 1);
    // If high-byte escape tag use the following byte plus 1
    else if (b === 0xfe) bytes.push(buffer[++i] + 1);
    // Otherwise no unescapement needed
    else bytes.push(b);
  }
  return new Buffer(bytes);
}


function parseHead(buffer) {
  // Parses and returns the first type on the buffer and the total bytes consumed
  var type = buffer[0];
  // Nullary
  if (~nullaryTypes.indexOf(type)) return [ decode(new Buffer([ type ])), 1 ];
  // Fixed
  var size = fixedTypes[type];
  if (size++) return [ decode(buffer.slice(0, size)), size ];
  // Flat
  var index;
  var end;
  if (~flatTypes.indexOf(type)) {
    // Find end byte
    for (index = 1, end = buffer.length; index < end; ++index) {
      if (buffer[index] === 0x00) break;
    }
    if (index >= buffer.length) throw new Error('No ending byte found for list');
    var chunk = flatUnescape(buffer.slice(0, index));
    // Add 1 to index to skip over end byte
    return [ decode(chunk), index + 1 ];
  }
  // Nested, recurse for each item
  var list = [];
  index = 1;
  var next;
  while ((next = buffer[index]) !== 0) {
    var result = parseHead(buffer.slice(index));
    list.push(result[0]);
    index += result[1];
    if (index >= buffer.length) throw new Error('No ending byte found for nested list');
  }
  return [ structure(type, list), index + 1 ];
}

function structure(type, list) {
  if (type === ARRAY) return list;
  if (type === FUNCTION) {
    return _type['function'].parse(list);
  }
  var i, end;
  if (type === OBJECT) {
    var object = Object.create(null);
    for (i = 0, end = list.length; i < end; ++i) {
      object[list[i]] = list[++i];
    }
    return object;
  }
  if (type === MAP) {
    var map = new Map();
    for (i = 0, end = list.length; i < end; ++i) {
      map.set(list[i], list[++i]);
    }
    return map;
  }
  if (type === SET) {
    var set = new Set();
    for (i = 0, end = list.length; i < end; ++i) {
      set.add(list[i]);
    }
    return set;
  }
  throw new Error('Unknown type: ' + type);
}


function invert(buffer) {
  var bytes = [];
  for (var i = 0, end = buffer.length; i < end; ++i) {
    bytes.push(~buffer[i]);
  }
  return new Buffer(bytes);
}


function _bastardizedIterate(iterator) {
  // FIXME update es6-shim to latest iteration spec and remove this garbage
  var items = [];
  var next;
  try {
    while (next = iterator.next()) {
      items.push(next)
    }
  }
  catch (e) {}
  return items;
}

function getCollectionKeys(collection) {
  // FIXME support --harmony collection iteration
  if (!typeof collection.keys === 'function') return [];
  return _bastardizedIterate(collection.keys());
}


exports.encode = encode;
exports.decode = decode;
exports.compare = compare;
